python/tvm/build_module.py

```diff
@@ -167,6 +171,37 @@ def get_binds(args, binds=None):
             raise ValueError("args must be Tensor, Buffer or Var")
     return binds, arg_list
 
+def dump_ir(debug, pass_inp=None):
+    ''' dump ir for each pass, includes costomized pass'''
+    def decorator(func, pass_id=[]): # pylint: disable=dangerous-default-value
+        ''' decorate the pass function, use pass_id[0] as increased pass id'''
+        def dump(*args, **kwargs):
+            '''dump function'''
+            retv = func(*args, **kwargs)
+            if isinstance(retv, expr.IntImm):
+                return retv
+            pname = str(pass_id[0]) + "_" + func.func_name + "_ir.cc"
+            with open(pname, "w") as f:
+                print >> f, (retv.body if isinstance(retv, container.LoweredFunc) else retv)
+                if func.func_name == "SplitHostDevice":
+                    for x in retv:
+                        print >> f, "---------", x.name, "\n", x.body, "\n-----------\n"
+                pass_id[0] += 1
+            return retv
+        if not pass_id:
+            pass_id.append(0)
+        return dump
+
+    def decorator_passes(add_lower_pass):
+        '''decorate ir_pass and add_lower_pass'''
+        for k, v in vars(ir_pass).items():
+            vars(ir_pass)[k] = decorator(v) if isinstance(v, types.FunctionType) else v
+        schedule.ScheduleOps = decorator(schedule.ScheduleOps)
```

Comment:
should not decorate a function big not recover it later. A general decorator pattern should be something like
```
with decorate(list_of_functions):
     things take in effect
recover decoration to the old-list of functions
```

Reply:
1) since passes are called in lower and build, so this with should like buid_config, in which __enter and __exit are defined;
2) since ir_pass is a module, as far as i know, the way to recover it is reload(ir_pass), and i'm not sure it's a proper way to do.
I'm going to refactor it as build_config